File: 278-firstBadVersion.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        a = 0
        b = n
        while b - a != 1:
            if isBadVersion((a + b) // 2) == True:
                b = (a + b) // 2
            else:
                a = (a + b) // 2
        return b

File: 278-firstBadVersion.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int a = 0;
        int b = n;
        int mid;
        while (a < b){
            mid = a/2 + b/2; //使用(a+b)/2会有溢出
            if (isBadVersion(mid)){
                b = mid;
            }else {
                a = mid + 1;
            }
        }
        return a;
    }
}